streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.kafka.streams.state.internals.ThreadCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * This is the thread responsible for keeping all Global State Stores updated.
+ * It delegates most of the responsibility to the internal class StateConsumer
+ */
+public class GlobalStreamThread extends Thread {
+
+    private static final Logger log = LoggerFactory.getLogger(GlobalStreamThread.class);
+    private final StreamsConfig config;
+    private final Consumer<byte[], byte[]> consumer;
+    private final StateDirectory stateDirectory;
+    private final Time time;
+    private final ThreadCache cache;
+    private final StreamsMetrics streamsMetrics;
+    private final ProcessorTopology topology;
+    private volatile boolean running = false;
+    private volatile StreamsException startupException;
+
+    public GlobalStreamThread(final ProcessorTopology topology,
+                              final StreamsConfig config,
+                              final Consumer<byte[], byte[]> globalConsumer,
+                              final StateDirectory stateDirectory,
+                              final Metrics metrics,
+                              final Time time,
+                              final String clientId
+    ) {
+        super("GlobalStreamThread");
+        this.topology = topology;
+        this.config = config;
+        this.consumer = globalConsumer;
+        this.stateDirectory = stateDirectory;
+        this.time = time;
+        long cacheSizeBytes = Math.max(0, config.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG) /
+                (config.getInt(StreamsConfig.NUM_STREAM_THREADS_CONFIG) + 1));
+        final String threadClientId = clientId + "-" + getName();
+        this.cache = new ThreadCache(threadClientId, cacheSizeBytes, new ThreadCache.NullThreadCacheMetrics());
+        this.streamsMetrics = new StreamThread.StreamsMetricsImpl(metrics, threadClientId);
+    }
+
+    static class StateConsumer {
+        private final Consumer<byte[], byte[]> consumer;
+        private final GlobalStateMaintainer stateMaintainer;
+        private final Time time;
+        private final long pollMs;
+        private final long flushInterval;
+
+        private long lastFlush;
+
+        StateConsumer(final Consumer<byte[], byte[]> consumer,
+                      final GlobalStateMaintainer stateMaintainer,
+                      final Time time,
+                      final long pollMs,
+                      final long flushInterval) {
+            this.consumer = consumer;
+            this.stateMaintainer = stateMaintainer;
+            this.time = time;
+            this.pollMs = pollMs;
+            this.flushInterval = flushInterval;
+        }
+
+        void initialize() {
+            final Map<TopicPartition, Long> partitionOffsets = stateMaintainer.initialize();
+            consumer.assign(partitionOffsets.keySet());
```

Comment:
This consumer is not the restore consumer. It is a consumer just for this thread so it isn't shared at all. There are no parallel operations on it.
When the `StateStores` are restored, in `GlobalStateManagerImpl`, they each `assign` their own partitions, fetch the data up to the HW, then un-assign their partitions. During this process all of the `TopicPartition` for global stores are collected and that is what is assigned here. This is the set of partitions we need to consume to keep all global stores up-to-date.

Yes, recovery of StateStores is always single threaded.